relay-metrics/src/aggregation.rs

```diff
@@ -1729,11 +1753,57 @@ impl Aggregator {
         buckets
     }
 
+    /// Split buckets into N logical partitions, determined by the bucket key.
+    fn partition_buckets(
+        &self,
+        buckets: Vec<HashedBucket>,
+        flush_partitions: u32,
+    ) -> BTreeMap<u32, Vec<Bucket>> {
+        let mut partitions = BTreeMap::<u32, Vec<Bucket>>::new();
+        for bucket in buckets {
+            let partition_key = bucket.hashed_key % flush_partitions;
+            partitions
+                .entry(partition_key)
+                .or_default()
+                .push(bucket.bucket);
+        }
+        partitions
+    }
+
+    /// Split the provided buckets into batches and process each batch with the given function.
+    ///
+    /// For each batch, log a histogram metric.
+    /// NOTE: This function can be inlined again once we are done with the dry run.
+    fn process_batches<F>(
+        &self,
+        buckets: impl IntoIterator<Item = Bucket>,
+        partition_key: Option<u32>,
+        mut process: F,
+    ) where
+        F: FnMut(Vec<Bucket>),
+    {
+        let capped_batches =
+            CappedBucketIter::new(buckets.into_iter(), self.config.max_flush_bytes);
+        let partition_tag = match partition_key {
+            Some(partition_key) => format!("{partition_key}"),
+            None => "none".to_owned(),
+        };
+        let capped_batches: Vec<_> = capped_batches.collect();
+        for (i, batch) in capped_batches.into_iter().enumerate() {
+            relay_statsd::metric!(
+                histogram(MetricHistograms::BucketsPerBatch) = batch.len() as f64,
+                partition_key = partition_tag.as_str(),
+                batch_index = format!("{i}").as_str(),
```

Comment:
please don't tag this, as the cardinality is theoretically unbound, and instead add a histogram metric counting the number of batches within a partition, i.e. `metric!(histogram(..) = capped_batches.len());`